### ISSUE TITLE
【论文复现】定位到Dataloader有内存泄漏现象

### ISSUE THREAD

**Issue by CuberrChen** [PaddlePaddle/Paddle#35749]
- 标题：定位到Dataloader有内存泄漏现象
- 版本、环境信息：
    1）PaddlePaddle版本：2.1.2
    2）CPU：Aistudio 32G
    3）GPU：Aistudio Tesla 16G
    4）系统环境：ubuntu python3.7

- 训练信息
    1）单机，单卡
    2）显存信息 16G
    3）Operator信息
- 复现信息：Dataloader存在不稳定的内存泄漏，造成CPU内存缓慢增长直到最后内存溢出崩溃。
- 问题描述：
我利用memory_profiler工具对我的训练函数进行了profile分析：
```
@profile(precision=4,stream=open('memory_profiler.log','w+')) #only for debug
def memory_check(i, example,net)：
"""训练函数"""
```

从日志结果看，memory_check函数并没有造成内存的增加。但偶尔经历以下一次循环后就会出现CPU内存的增长。少则零点几M，多则几十M。因此我怀疑是dataloader的内存泄漏。
```
for i, example in enumerate(dataloader):
      memory_check(i,example,net)
```

- 截取的日志：
**前一次memory_check的profile日志结尾**：

|Line|Mem usage|Increment|Occurences|Line Contents|
 |---|---|---|---|---|                                         
  |493| 3128.1992 MiB  | 0.0000 MiB |         1  |     return|

**后一次memory_check的profile日志开头**：
Filename: train_mgpu.py

|Line|Mem usage|Increment|Occurences|Line Contents|
 |---|---|---|---|---|            
   |384 |3137.6602 MiB |3137.6602 MiB |          1|   @profile(precision=4,stream=open('memory_profiler.log','w+')) # for debug|
   |385| - | - |  - |                              def memory_check(i, example,net):|

可以看到相邻两次Dataloder数据读取，cpu内存增加了10M。

- Dataloder的定义方式：
```
 batch_sampler = paddle.io.DistributedBatchSampler(dataset, batch_size=input_cfg.BATCH_SIZE, shuffle=True, drop_last=True)
    dataloader = paddle.io.DataLoader(
        dataset=dataset,
        batch_sampler=batch_sampler,
        use_shared_memory=False,
        num_workers=input_cfg.NUM_WORKERS,
        return_list=True,
        collate_fn=merge_second_batch,
        worker_init_fn=worker_init_fn
        )
```

请问该怎么解决，长时间训练后，会造成CPU内存溢出崩溃。

**Comment by ForFishes** [comment]
您好，这个可能是每次dataloder读取数据后，没有去覆盖之前读取的数据，导致泄漏。

**Comment by CuberrChen** [comment]
@ForFishes 但是并不是每次iter都会增加。是偶发且持续性增加。大概10个小时左右会把内存撑满。 那我可不可以手动再每一次iter结尾通过del指令手动清除数据来解决？

**Comment by CuberrChen** [comment]
通过每次iter结束处使用del 删除example并没有解决问题。

**Comment by Shawn6688** [comment]
我也遇到类似问题，内存缓慢增长，半天甚至一天后内存占满

**Comment by CuberrChen** [comment]
经过多次实验，确定就是

```
for i, example in enumerate(dataloader):
```

这一行代码造成的内存增加。 我想知道可能的原因，help！

- 第一次：
- 
```
   195 2985.0039 MiB 443.4258 MiB         102               for i, example in enumerate(dataloader):
   196                                                         
   197 2985.0039 MiB   0.0000 MiB         102                   if(i>100): #DEBUG
   198 2985.0039 MiB   0.0000 MiB           1                       break
```
batchsize为2，线程数为4，dataloader 100个iter后内存增加了443M。

- 第二次：
- 
```
   195 2531.4844 MiB   0.0000 MiB           1               i = 0
   196 2989.2500 MiB 297.0000 MiB         101               for example in dataloader:
   197 2989.2500 MiB -149.2227 MiB         101                   i += 1
   198 2989.2500 MiB -149.2227 MiB         101                   if(i>100): #DEBUG
   199 2989.2500 MiB   0.0000 MiB           1                       break
```

batchsize为2，线程数为4，dataloader 100个iter后内存增加了297M。


并且数据集的定义完全遵守paddle.io.Dataset的要求。

**Comment by tughv** [comment]
我用mnist数据集都能抛出线程错误，dataloader这个问题真头大

**Comment by ForFishes** [comment]
> 我用mnist数据集都能抛出线程错误，dataloader这个问题真头大

**Comment by ForFishes** [comment]
> 经过多次实验，确定就是
> 
> ```
> for i, example in enumerate(dataloader):
> ```
> 
> 这一行代码造成的内存增加。 我想知道可能的原因，help！
> 
> * 第一次：
> 
> ```
>    195 2985.0039 MiB 443.4258 MiB         102               for i, example in enumerate(dataloader):
>    196                                                         
>    197 2985.0039 MiB   0.0000 MiB         102                   if(i>100): #DEBUG
>    198 2985.0039 MiB   0.0000 MiB           1                       break
> ```
> 
> batchsize为2，线程数为4，dataloader 100个iter后内存增加了443M。
> 
> * 第二次：
> 
> ```
>    195 2531.4844 MiB   0.0000 MiB           1               i = 0
>    196 2989.2500 MiB 297.0000 MiB         101               for example in dataloader:
>    197 2989.2500 MiB -149.2227 MiB         101                   i += 1
>    198 2989.2500 MiB -149.2227 MiB         101                   if(i>100): #DEBUG
>    199 2989.2500 MiB   0.0000 MiB           1                       break
> ```
> 
> batchsize为2，线程数为4，dataloader 100个iter后内存增加了297M。
> 
> 并且数据集的定义完全遵守paddle.io.Dataset的要求。

您好，这个dataset问题，会反馈给相关同学帮忙看下。

**Comment by heavengate** [comment]
2.1.1版本上DataLoader有个线程和内存不能释放的泄漏问题，2.1.2版本已经修复了，可以更新下Paddle版本试下

**Comment by CuberrChen** [comment]
@heavengate 好的 我换下版本试下

**Comment by tughv** [comment]
我的就是2.1.2，还是使用的官方mnist数据集
`import paddle 
from paddle.vision.models import LeNet
from paddle.vision.transforms import Compose,Normalize,Grayscale,Resize
from paddle.metric import Accuracy
import paddle.nn.functional as F
#使用transform对数据集做归一化
transform = Compose([Normalize(mean=(127.5),std=[127.5],data_format="CHW"),Grayscale(num_output_channels=1),Resize(28)])
train_dataset = paddle.vision.datasets.MNIST(mode="train",transform=transform,download=True)
test_dataset = paddle.vision.datasets.MNIST(mode="test",transform=transform,download=True)

# model = paddle.Model(LeNet())
model = LeNet()

train_loader = paddle.io.DataLoader(train_dataset,batch_size=16,shuffle=True,use_shared_memory=False)
paddle.device.set_device("gpu")
def train(model,epochs):
    model.train()
    optim = paddle.optimizer.Adam(learning_rate=0.001,parameters=model.parameters())
    for epoch in range(epochs):
        for batch_id, (img,label) in enumerate(train_loader):
            print(img.shape)
            pred = model(img)
            loss = F.CrossEntropyLoss(pred,label)
            acc = paddle.metric.accuracy(pred, label)
            loss.backward()
            if batch_id%300==0:
                print("epoch: {}, batch_id: {}, loss is: {}, acc is: {}".format(epoch, batch_id, loss.numpy(), acc.numpy()))
            optim.step()
            optim.clear_grad()
model=LeNet()
train(model,epochs=20)`
[
![image](https://user-images.githubusercontent.com/40344969/133708931-926923d2-555b-4157-960a-718994835fbe.png)
](url)

**Comment by CuberrChen** [comment]
我用的就是2.1.2 。还是找不到问题所在。我检查了数据预处理代码，全是numpy的数据处理。

**Comment by Shawn6688** [comment]
我也是2.1.2

**Comment by MarginGitHub** [comment]
我也是，每次训练10个小时左右必然耗尽内存，然后导致训练被killed，我的训练机器内存是128G的，paddle版本是2.1.2

**Comment by tughv** [comment]
我启动就报错，一轮训练都没有开始，使用torch去实现一样的就没有问题

**Comment by CuberrChen** [comment]
问题未解决，使用了另外一种训练方法：收集内存信息，大于一定阈值终止训练等到内存释放后再继续resume。太难了..
无论最后比赛结束是否达到题目验收标准，都会开源项目，希望到时感兴趣的大佬/飞浆工作人员看下哪里有问题。

**Comment by Shawn6688** [comment]
我自己写了dataloader，发现并不是官方的dataloader导致的内存泄漏。进行了一些实验目前追踪到是paddle.jit.save()保存模型和保存loss.numpy()都会引起内存泄露。怀疑是tensor没有从计算图分离导致的。如果有新结果再反馈。

**Comment by chenkangyang** [comment]
追更~ 同样遇到内存泄漏的问题~ 不仅dataloader，模型推理算loss也泄漏了；

**Comment by chenkangyang** [comment]
> 我自己写了dataloader，发现并不是官方的dataloader导致的内存泄漏。进行了一些实验目前追踪到是paddle.jit.save()保存模型和保存loss.numpy()都会引起内存泄露。怀疑是tensor没有从计算图分离导致的。如果有新结果再反馈。

相邻两个batch，都不需要算loss，推理一次，内存就会增加一点
```
Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
   892   3551.2 MiB   3551.2 MiB           1   @profile
   893                                         def run_on_batch(model, return_dict, inputs, labels, t_loss, optimizer, lr_scheduler):
   894                                             
   895   3553.6 MiB      2.4 MiB           1       outputs = model(**inputs)


Filename: check_dataloader_just_img.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
   892   3553.6 MiB   3553.6 MiB           1   @profile
   893                                         def run_on_batch(model, return_dict, inputs, labels, t_loss, optimizer, lr_scheduler):
   894                                             
   895   3555.1 MiB      1.6 MiB           1       outputs = model(**inputs)
```

**Comment by Shawn6688** [comment]
你这个outputs =
model(**inputs)后边还有代码吗？你可以尝试把后边的注释掉只保留到这样一行，看会不会泄露。我注释掉model()后的代码就没有泄露了。
他确实在batch这一行内存增加，但是根源可能在其他地方产生的计算图依赖。

**Comment by CuberrChen** [comment]
@Shawn6688 实验过了。只dataloader这一行，下面空循环也会内存增长。早知道不如自己写个loader了。查半天错。不过暂时没有余力改了，先这样吧。谢谢～

**Comment by CuberrChen** [comment]
已经公开项目：https://github.com/CuberrChen/Voxelnet-Paddle
AiStudio链接：https://aistudio.baidu.com/aistudio/projectdetail/2312826?contributionType=1
希望大佬可以看下哪里有问题。
